pkg/cassandra/nodetool/nodetool.go

```diff
@@ -0,0 +1,170 @@
+
+func (nm NodeMap) LocalNode() *Node {
+	for _, node := range nm {
+		if node.Local {
+			return node
+		}
+	}
+	return nil
+}
+
+type Interface interface {
+	Status() (NodeMap, error)
+}
+
+type tool struct {
+	client client.Interface
+}
+
+var _ Interface = &tool{}
+
+func New(client client.Interface) Interface {
+	return &tool{
+		client: client,
+	}
+}
+
+func NewFromURL(u *url.URL) Interface {
+	return New(client.New(u, &http.Client{}))
+}
+
+func setsIntersect(setsToCheck ...sets.String) bool {
+	all := sets.NewString()
+	totalLength := 0
+	for _, s := range setsToCheck {
+		totalLength += s.Len()
+		all = all.Union(s)
+	}
+	return all.Len() != totalLength
+}
+
+func (t *tool) Status() (NodeMap, error) {
+	ssInfo, err := t.client.StorageService()
+	if err != nil {
+		return nil, err
+	}
+
+	nodes := NodeMap{}
+	mappedNodes := sets.NewString()
+	for host, id := range ssInfo.HostIdMap {
+		mappedNodes.Insert(host)
+		nodes[host] = &Node{
+			Host:   host,
+			ID:     id,
+			Status: NodeStatusUnknown,
+			State:  NodeStateNormal,
+		}
+	}
+
+	liveNodes := sets.NewString(ssInfo.LiveNodes...)
+	unreachableNodes := sets.NewString(ssInfo.UnreachableNodes...)
+	if setsIntersect(liveNodes, unreachableNodes) {
+		return nil, fmt.Errorf(
+			"The sets of live and unreachable nodes should not intersect. "+
+				"Live: '%s', "+
+				"Unreachable: '%s'",
+			liveNodes, unreachableNodes,
+		)
+	}
+	if !mappedNodes.IsSuperset(liveNodes.Union(unreachableNodes)) {
+		return nil, fmt.Errorf(
+			"Mapped nodes must be a superset of Live and Unreachable nodes. "+
+				"Live: '%s', "+
+				"Unreachable: '%s', "+
+				"Mapped: '%s'",
+			liveNodes, unreachableNodes, mappedNodes,
+		)
+	}
+
+	leavingNodes := sets.NewString(ssInfo.LeavingNodes...)
+	joiningNodes := sets.NewString(ssInfo.JoiningNodes...)
+	movingNodes := sets.NewString(ssInfo.MovingNodes...)
+
+	if setsIntersect(leavingNodes, joiningNodes, movingNodes) {
+		return nil, fmt.Errorf(
+			"The sets of leaving, joining and moving nodes should not intersect. "+
+				"Leaving: '%s', "+
+				"Joining: '%s', "+
+				"Moving: '%s'",
+			leavingNodes, joiningNodes, movingNodes,
+		)
+	}
+
+	if !mappedNodes.IsSuperset(leavingNodes.Union(joiningNodes).Union(movingNodes)) {
+		return nil, fmt.Errorf(
+			"Mapped nodes must be a superset of leaving, joining and moving nodes. "+
+				"Leaving: '%s', "+
+				"Joining: '%s', "+
+				"Moving: '%s', "+
+				"Mapped: '%s'",
+			leavingNodes, joiningNodes, movingNodes, mappedNodes,
+		)
+	}
+
+	for host, node := range nodes {
+		switch {
+		case liveNodes.Has(host):
+			node.Status = NodeStatusUp
+		case unreachableNodes.Has(host):
+			node.Status = NodeStatusDown
+		default:
+			node.Status = NodeStatusUnknown
+		}
+
+		switch {
+		case leavingNodes.Has(host):
+			node.State = NodeStateLeaving
+		case joiningNodes.Has(host):
+			node.State = NodeStateJoining
+		case movingNodes.Has(host):
+			node.State = NodeStateMoving
+		default:
+			node.State = NodeStateNormal
```

Comment:
Is there ever any case where NodeStateUnknown would be returned? Can't see it here.

Reply:
Yeah, if a node isn't in either of LiveNodes or Unreachable nodes.
The `?` case here: https://git1-us-west.apache.org/repos/asf?p=cassandra.git;a=blob;f=src/java/org/apache/cassandra/tools/nodetool/Status.java;h=f7bb1e52a786fe40ec9a2d6d866618c814562b3f;hb=refs/heads/cassandra-3.11#l155

Reply:
So I see we have `NodeStatusUnknown` as well as `NodeStateUnknown`, although `NodeStateUnknown` is not used in this `for` loop, hence my question (as it seems that there are no cases where `node.State` is set to `NodeStateUnknown`).

Reply:
Yeah from re-reading this whole function it appears we should set `node.State = NodeStateNormal` *iff* the node is a member of `liveNodes`? Otherwise we should set this to unknown (which seems equiv. to 'unreachable'? Unless we need that as an *extra* state).

Reply:
I'm pretty sure I'm exactly replicating the logic from nodetool.

```
 153         if (liveNodes.contains(endpoint)) status = "U";
 154         else if (unreachableNodes.contains(endpoint)) status = "D";
 155         else status = "?";
 156         if (joiningNodes.contains(endpoint)) state = "J";
 157         else if (leavingNodes.contains(endpoint)) state = "L";
 158         else if (movingNodes.contains(endpoint)) state = "M";
 159         else state = "N";
```

The reason for `NodeStateUnknown` is that that I wanted a sensible zero-value.
You're right that the `nodetool.Status` function will never return it, but none of the other states make sense as zero value.